### ISSUE TITLE
data: concurrent transforms and prefetch through a C++ worker pool

### ISSUE THREAD

**Issue by bkc39** [bkc39/rktorch#134]
Follow-up to #87 / PR #133. The loader there is synchronous (`num_workers=0`) on purpose: with a device-resident `tensor-dataset` a batch is a `narrow` view or one `index-select`, and on MNIST the loader's own share of an epoch is 11 ms sequential, 43 ms shuffled, out of ~1750 ms on the 3090 Ti. Nothing to overlap.

The customer for a concurrent loader is per-sample **transforms and augmentations** (random crop with padding, horizontal flip, normalize, later random erasing / colour jitter) on data that is not already batched on the device, and the host-to-device copy behind them. That work must run in C++: a worker thread cannot call a Racket `dataset-ref`, and batch handles must cross the FFI on the Racket thread so the memory ledger never sees concurrent traffic (settled on #87).

## Proposed Racket surface

```racket
(define aug
  (compose-transforms (random-crop 32 #:padding 4)
                      (random-horizontal-flip)
                      (normalize mean std)))
(dataloader (tensor-dataset xs ys)
            #:batch-size 128 #:shuffle? #t #:generator (make-generator 0)
            #:transform aug #:workers 4 #:prefetch 2)
```

- `#:transform` applies to field 0 per sample, before collate, as a `Dataset.__getitem__` transform does.
- `#:workers 0` (default) applies a transform on the calling thread, drawing from the global stream, as `num_workers=0` does. Any Racket procedure is allowed here.
- `#:workers > 0` requires a CPU `tensor-dataset` and a native transform object (contract). Batches come back CPU-resident; pinned memory and an asynchronous device copy are a second step.

## Proposed C API (`c_api/data.h`)

```c
tr_transform* tr_transform_random_crop(int64_t h, int64_t w, int64_t padding, double fill);
tr_transform* tr_transform_random_hflip(double p);
tr_transform* tr_transform_normalize(const double* mean, const double* std, int64_t n);
tr_transform* tr_transform_compose(tr_transform* const* ts, int64_t n);
void          tr_transform_free(tr_transform*);
tr_tensor*    tr_transform_apply(tr_transform*, tr_tensor* x, tr_generator* g_or_null); /* one sample, synchronous */

tr_loader* tr_loader_new(tr_tensor* const* fields, int64_t nfields, int64_t batch_size,
                         int drop_last, tr_transform* xf, int64_t workers, int64_t prefetch);
int  tr_loader_start(tr_loader*, tr_tensor* perm_or_null, uint64_t base_seed);
int  tr_loader_next(tr_loader*, tr_tensor** out /* nfields */);   /* 0 batch, 1 exhausted, 2 error */
void tr_loader_free(tr_loader*);                                   /* joins the pool */
```

The sampler draws (base seed, permutation, discarded trailing permutation) stay on the Racket side as in #133; the base seed is passed down for worker seeding.

## Parity: exact replay of `DataLoader(num_workers=W)` (owner's choice)

- Worker `w` owns a CPU generator seeded `base_seed + w`, as PyTorch seeds each worker process.
- Batch `k` goes to worker `k mod W`; at most `prefetch * W` batches are in flight; results are delivered in order. That is PyTorch's round-robin assignment with `prefetch_factor`, so a worker's RNG consumption depends only on the batches it processed before.
- Each transform draws exactly what torchvision draws: `RandomCrop` two `randint`s (none when the padded image equals the crop), `RandomHorizontalFlip` one `rand`, `Normalize` none.
- A Python twin (`Dataset` applying `transforms.v2` in `__getitem__`, `DataLoader(num_workers=W, generator=g)`) pins batches bit-for-bit for `W = 0` and `W = 2`.
- Prerequisite already in #133: every epoch draws the base seed, shuffled or not, because PyTorch creates it for every iterator.

## Design notes

- Errors: the error buffer is thread-local, so a worker's failure is captured in its slot and surfaced by `tr_loader_next` on the calling thread with the worker's message.
- `tr_loader_next` blocks the calling OS thread while a batch is pending, as `next(iter(loader))` does; Racket green threads pause with it. Acceptable for a training loop; a polling variant can come later if a caller needs it.
- `tr_loader_free` joins the pool; the Racket finalizer path goes through `swallow-and-count-failure` like every other handle.
- Not `torch::data::DataLoader`: its sampler and transform types do not reproduce PyTorch's per-worker RNG scheme, and the ordering guarantee is easier to state in a small pool of our own.

## Legs

1. Transforms and `tr_transform_apply`, gtests, single-sample torchvision twin.
2. Worker pool, gtests for ordering, replay, worker error surfacing, and join on free.
3. `torch/data/transforms.rkt`, loader keywords, docs chapter, `num_workers=2` twin.
4. Benchmark `#:workers 0` against `4` on `[N 3 32 32]` data with the CIFAR augmentation pipeline; then adopt it in the diffusion epic (#84) if it pays.

Refs: #40 (thread-safety audit), #87, #84.
